File: src/ba03/ba03i_k_universal/ba03i.py
"""
Rosalind: BA3I
Find a k-Universal Circular String

A k-universal circular string is a circular string
that contains every possible k-mer constructed over a given alphabet.

k-Universal Circular String Problem
Find a k-universal circular binary string.

Given: An integer k.

Return: A k-universal circular string.
(If multiple answers exist, you may return any one.)

Sample Dataset
4

Sample Output
0000110010111101

═════════════════════════════════════════════════

    [ Where am I? ]

    * How do we assemble genome?
      ↓
    * String reconstruction problem
      -> Generate the k-mer Composition of a String (BA3A)
      -> String Reconstruction Problem with k-mer composition(BA3B)
      ↓
    * Construct OverapGraph (BA3C)
      -> Reconstruct a String from its k-mer Composition (BA3H)
      ↓
    * De Bruijn Graph
      -> Construct De Bruijn Graph (BA3D)
      -> Construct De Bruijn Graph with k-mers (BA3E)
      ↓
    * Eulerian Cycle, Path
      -> Eulerian Cycle (BA3F)
        -> HERE: k-universal string problem (BA3I)
      -> PREV: Eulerian Path (BA3G)
      ↓
    * Read-Pair (BA3J)
      -> NEXT: Construct a String Spelled by a Gapped Genome Path (BA3L)


Plan 1.
* steps:
  i.   create a list k-mers - cartesian product with "01"
       All (not some) of these binary k-mers should be used to reconstruct k-universal string.
  ii.  get De Bruijn graph from the list (exclude edges to node itself)
       split each k-mer into prefix and suffix (k-1)-mers
       then construct De Bruijn graph represented as dictionary
  iii. get Eulerian cycle from De Bruijn graph
  iv.  find out the way how to reconstruct string from Eulerian cycle
       BA3B string reconstruction function is not for circular. Be careful!
       ['00','01','11','11','10','01','10','00','00']

       index     0      1 ...     -3      -2      -1
       ---------------------------------------------
       k=2       0      1          1       0   /   0
       k=3      00     01 ...     10   /  00      00
       k=4     000    001 ...   10/0     000     000
       k=5    0000   0001 ...  10/00    0000    0000
       k=6   00000  00001 ... 10/000   00000   00000
                              -- ---
                             (a)  (b)   <-- (a) is always '10', but (b) changes

═════════════════════════════════════════════════

References:
- fast factorial problem
  - Why python math.factorial(x) is very fast?
    https://stackoverflow.com/questions/11313998/why-python-math-factorialx-is-very-fast
  - Tabulation vs Memoization
    https://www.geeksforgeeks.org/tabulation-vs-memoization/
  - Dynamic Programming Tutorial: making efficient programs in Python
    https://www.educative.io/blog/python-dynamic-programming-tutorial
- Generate all permutation of a set in Python
  https://www.geeksforgeeks.org/generate-all-the-permutation-of-a-list-in-python/
- How to generate all permutations of a list?
  https://stackoverflow.com/questions/104420/how-to-generate-all-permutations-of-a-list
- How to print like printf in Python3?
  https://stackoverflow.com/questions/19457227/how-to-print-like-printf-in-python3
"""
#!/usr/bin/env python
import time
import itertools


def prod_itertools(text, k):
    """
    (str,int) -> [str]
    >>> prod_itertools("ACG",2)
        ['AA','AC','AG','CA','CC','CG','GA','GC','GG']
    """
    from itertools import product
    result = []
    for tup in itertools.product(text, repeat=k):
        temp = ''
        for elm in tup:
            temp += elm
        result.append(temp)
    return result


def prod_recur(text, k):
    """
    (str,int) -> gen
    >>> list(prod_recur("ACG",2))
        ['AA','AC','AG','CA','CC','CG','GA','GC','GG']
    """
    if k == 1:
        for ch in text:
            yield ch
    else:
        for ch in text:
            for s in prod_recur(text, k - 1):
                yield ch + s


def prod_lscomp(text, k):
    """
    (str,int) -> [str]
    >>> prod_lscomp('01',3)
        ['000','001','010','011','100','101','110','111']
    """
    return ["".join(a) for a in itertools.product(text, repeat = k)]


def de_bruijn(kmers):
    """
    [a] -> {a:[a]}
    returns DeBruijnk graph from k-mer patterns (BA3E)
    representation of De Bruijn graph as list of tuples: not good idea, use dictionary
    >>> de_bruijn(['GAGG','CAGG','GGGG','GGGA','CAGG','AGGG','GGAG'])
        {'GAG':['AGG'],'CAG':['AGG','AGG'],'GGG':['GGG','GGA'],'AGG':['GGG'],'GGA':['GAG']}
    >>> de_bruijn( ['000','001','010','011','100','101','110','111'])
        {'00':['00','01'],'01':['10','11'],'10':['00','01'],'11':['10','11']}
    """
    result = {}
    for kmer in kmers:
        key = kmer[:-1]
        val = kmer[1:]
        if key not in result:
            result[key] = [val]
        else:
            result[key].append(val)
    return result


def eulerian_cycle(graph):
    """
    {a:[a]} -> [a]
    returns Eulerian cycle
    """
    cycle = [min(graph.keys())]
    while len(graph) > 0:
        if cycle[0] == cycle[-1]:
            while not cycle[0] in graph:
                cycle.pop(0)
                cycle.append(cycle[0])
        source = cycle[-1]
        cycle.append(graph[source].pop())
        if len(graph[source]) == 0: del graph[source]
    return cycle


def str_reconst(kmers):
    """
    [str] -> str
    returns a merged string from overlapping k-mers (BA3B)
    >>> str_reconst(['00','01','11','11','10','01','10','00','00'])
        00011101
    """
    result = kmers[0]
    for node in kmers[1:]:
        result += node[-1]
    return result


def k_universal_strings(k):
    """
    int -> {str}
    returns k-universal circular string
    """
    kmers = prod_lscomp('01',k)
    graph = de_bruijn(kmers)
    cycle = eulerian_cycle(graph)
    indices = [i for i, x in enumerate(cycle) if x == cycle[0]]
    result = ""
    if k > 3:
        result = str_reconst(cycle[:-2])[:-(k-3)]
    return result


def main():
    f = open('/home/wsl/rosalind/data/ba03i.txt', 'r')
    k = int(f.readline().strip())
    f.close()

    start_time = time.time()
    print(k_universal_strings(k))
    print("--- %s seconds ---" % (time.time() - start_time))


if __name__ == "__main__":
    main()

File: src/ba03/ba03i_k_universal/ba03i.js
/*
Rosalind: BA3I
Find a k-Universal Circular String

A k-universal circular string is a circular string
that contains every possible k-mer constructed over a given alphabet.

k-Universal Circular String Problem
Find a k-universal circular binary string.

Given: An integer k.

Return: A k-universal circular string.
(If multiple answers exist, you may return any one.)

Sample Dataset
4

Sample Output
0000110010111101

═════════════════════════════════════════════════

    [ Where am I? ]

    * How do we assemble genome?
      ↓
    * String reconstruction problem
      -> Generate the k-mer Composition of a String (BA3A)
      -> String Reconstruction Problem with k-mer composition(BA3B)
      ↓
    * Construct OverapGraph (BA3C)
      -> Reconstruct a String from its k-mer Composition (BA3H)
      ↓
    * De Bruijn Graph
      -> Construct De Bruijn Graph (BA3D)
      -> Construct De Bruijn Graph with k-mers (BA3E)
      ↓
    * Eulerian Cycle, Path
      -> Eulerian Cycle (BA3F)
        -> HERE: k-universal string problem (BA3I)
      -> PREV: Eulerian Path (BA3G)
      ↓
    * Read-Pair (BA3J)
      -> NEXT: Construct a String Spelled by a Gapped Genome Path (BA3L)

Info.
  * sample dataset if k=3, k-universal string is "01000111"
    01           010    .
     10          .100   .
      00         . 000  .
       00        .  001 .
        01       .   011.
         11      .    111 <--
          11     .     11-0
           10    .      1-01
            01
    0100011101   01000111-01
    if linear    if cycle

Plan 1.
  * product
    k=1 [0,1] 2^1  [0,1]
    k=2 [0,1] 2^2  [00,01,10,11]
    k=3 [0,1] 2^3  [000,001,010,100,011,101,110,111]

  * steps:
    i.   create a list k-mers - cartesian product with "01"
         All (not some) of these binary k-mers should be used to reconstruct k-universal string.
    ii.  get De Bruijn graph from the list (exclude edges to node itself)
         split each k-mer into prefix and suffix (k-1)-mers
         then construct De Bruijn graph represented as dictionary
    iii. get Eulerian cycle from De Bruijn graph
         TODO: Is the graph always Eulerian cycle? Always balanced?
               Strangely, the graph is always Eulerian. Why?
         TODO: If the graph is not Eulerian, what shall I do?
    iv.  find out the way how to reconstruct string from Eulerian cycle
         BA3B string reconstruction function is not for circular. Be careful!
         - temp_string = cycle[0] + last characters from cycle[1:]
         - final_string = temp_string[:-(k-1)]

═════════════════════════════════════════════════

References:
- Generate permutations of JavaScript array
  https://stackoverflow.com/questions/37579994/generate-permutations-of-javascript-array
*/

// helper fx. : get all product
function product_rec(arr, k) {
    /**
     * ([str],int) -> [str]
     * returns a list of product from fiven list of characters
     * >>> product_rec(['0','1'], 2)
     *     ['00','01','10','11]
     */
    if (k == 1) return arr;
    let res = [];
    for (const elem of arr) {
        for (const s of product_rec(arr, k - 1)) {
            res.push(elem + s);
        }
    }
    return res;
}

// helper fx. : get all product, generator version
function* product_rec_gen(arr, k) {
    /**
     * ([str],int) -> generator
     * returns a list of product from fiven list of characters, generator version
     * I asked stackoverflow about how to implement this function.
     * https://stackoverflow.com/questions/72195200/javascript-getting-all-possible-k-length-strings-from-given-list-of-characters
     */
    for (let i = 0; i < arr.length; i++) {
        if (k === 1) {
            yield [arr[i]];
        } else {
            const rest = product_rec_gen(arr, k - 1);
            for (let el of rest) {
                yield arr[i] + el;
            }
        }
    }
}

// BA03E: De Bruijn graph, from k-mers
function debruijn_from_kmers(kmers) {
    /**
     * [str] -> {str:[str]}
     * returns a De Bruijn graph
     */
    let graph = {};
    for (const kmer of kmers) {
        const prefix = kmer.slice(0, kmer.length - 1);
        const suffix = kmer.slice(1);
        if (!(prefix in graph)) graph[prefix] = [suffix];
        else graph[prefix].push(suffix);
    }
    return graph;
}

// get degrees
function get_degrees(graph) {
    /**
     * {str:[a]} -> {str:[int,int]}       // no tuple type in Javascript, object property is always string
     * returns In/Out-degrees
     * >>> get_degrees({ 0:[3],1:[0],2:[1,6],3:[2],4:[2],5:[4],6:[5,8],7:[9],8:[7],9:[6] })
     *     {'0':[1,1],'1':[1,1],'2':[2,2],'3':[1,1],'4':[1,1],'5':[1,1],'6':[2,2],'7':[1,1],'8':[1,1],'9':[1,1]}
     */
    let dic = {};
    for (const key in graph) {
        // count In-degrees
        for (const val of graph[key]) {
            if (!(val in dic)) dic[val] = [1, 0];
            else dic[val][0]++;
        }
        // count Out-degrees
        if (!(key in dic)) dic[key] = [0, graph[key].length];
        else dic[key][1] += graph[key].length;
    }
    return dic;
}

// check if graph is Eulerian
function check_eulerian_cycle(graph) {
    /**
     * {str:[a]} -> bool
     * returns true if graph is Eulerian cycle, else false (all nodes with even degrees)
     * >>> check_eulerian_cycle({0:[3],1:[0],2:[1,6],3:[2],4:[2],5:[4],6:[5,8],7:[9],8:[7],9:[6]})
     *     true
     */
    const degrees = get_degrees(graph);
    for (const key in degrees) {
        const sum = degrees[key].reduce((a, b) => a + b, 0);
        if (sum % 2 != 0) return false;
    }
    return true;
}

// simple walk
function walk(graph, node) {
    /**
     * ({str:[str]},str) -> [str]
     * returns a path or cycle by simple random walk
     */
    let cycle = [node];
    while (graph[node].length != 0) {
        const values = graph[node];
        const target = values[Math.floor(Math.random() * values.length)];
        cycle.push(target);
        graph[node] = values.filter(e => e != target);
        node = target;
    }
    return cycle;
}

// BA03F: Eulerian cycle, by Hierholzer's algorithm
function eulerian_hierholzer(graph) {
    /**
     * {a:[a]} -> [a]
     * returns a Eluerian cycle, destructive
     * >>> var graph = {'0':['3'],'1':['0'],'2':['1','6'],'3':['2'],'4':['2'],'5':['4'],'6':['5','8'],'7':['9'],'8':['7'],'9':['6']};
     * >>> eulerian_hierholzer(graph)
     *     ['6','5','4','2','1','0','3','2','6','8','7','9','6']
     */
    // first, get a cycle by random walk
    let nodes = Object.keys(graph);
    let source = nodes[Math.floor(Math.random() * nodes.length)];  // initial source node, randomly selected
    let cycle = walk(graph, source);
    // if there's remaining unvisited nodes, get another cycle, merge it to the previous cycle, again and again
    while (Object.keys(graph).map(e => graph[e].length === 0).every(Boolean) != true) {
        source = cycle.filter(e => graph[e].length != 0)[0];
        const new_cycle = walk(graph, source);
        const idx = cycle.indexOf(source);
        cycle = [...cycle.slice(0, idx + 1), ...new_cycle.slice(1), ...cycle.slice(idx + 1)];
    }
    return cycle;
}

// BA03I: k universal circular string
function k_universal_circular_string(k) {
    /**
     * int -> str
     * returns a list of k_universal circular strings
     */
    const all_kmers = product_rec(['0', '1'], k);
    const graph = debruijn_from_kmers(all_kmers);
    if (!(check_eulerian_cycle(graph))) return "Not Eulerian!"
    const cycle = eulerian_hierholzer(graph);
    let answer = cycle[0];
    for (const elem of cycle.slice(1)) {
        answer += elem[elem.length - 1];
    }
    answer = answer.slice(0, answer.length - (k - 1));
    return answer;
}

// main
function main() {
    const fs = require('fs');
    // const lines = fs.readFileSync('/home/wsl/rosalind/data/ba03i.txt').toString().split("\n");

    const startTime = performance.now()
    const cycle = k_universal_circular_string(4);
    console.log(cycle);
    console.log(`${performance.now() - startTime} milliseconds`)
}

// execute main function
main()
